src/backend/distributed/utils/reference_table_utils.c

```diff
@@ -95,56 +95,77 @@ EnsureReferenceTablesExistOnAllNodes(void)
 void
 EnsureReferenceTablesExistOnAllNodesExtended(char transferMode)
 {
+	List *referenceTableIdList = NIL;
+	uint64 shardId = INVALID_SHARD_ID;
+	List *newWorkersList = NIL;
+	const char *referenceTableName = NULL;
+	int colocationId = CreateReferenceTableColocationId();
+
 	/*
-	 * Prevent this function from running concurrently with itself.
-	 *
-	 * It also prevents concurrent DROP TABLE or DROP SCHEMA. We need this
-	 * because through-out this function we assume values in referenceTableIdList
-	 * are still valid.
+	 * Most of the time this function should result in a conclusion where we do not need
+	 * to copy any reference tables. To prevent excessive locking the majority of the time
+	 * we run our precondition checks first with a lower lock. If, after checking with the
+	 * lower lock, that we might need to copy reference tables we check with a more
+	 * aggressive and self conflicting lock. It is important to be self conflicting in the
+	 * second run to make sure that two concurrent calls to this routine will actually not
+	 * run concurrently after the initial check. That is also the reason why we release
+	 * the lock between the two iterations of precondition checks.
 	 *
-	 * We don't need to handle other kinds of reference table DML/DDL here, since
-	 * master_copy_shard_placement gets enough locks for that.
+	 * If after two iterations of precondition checks we still find the need for copying
+	 * reference tables we exit the forloop with the last lock held. This will prevent
+	 * concurrent DROP TABLE and create_reference_table calls so that the list of
+	 * reference tables we operate on are stable.
 	 *
-	 * We also don't need special handling for concurrent create_refernece_table.
-	 * Since that will trigger a call to this function from another backend,
-	 * which will block until our call is finished.
+	 * Since the changes to the reference table placements are made via loopback
+	 * connections we release the final lock held at the end of this function.
 	 */
-	int colocationId = CreateReferenceTableColocationId();
-	LockColocationId(colocationId, ExclusiveLock);
-
-	List *referenceTableIdList = CitusTableTypeIdList(REFERENCE_TABLE);
-	if (referenceTableIdList == NIL)
+	LOCKMODE lockmodes[] = { AccessShareLock, ExclusiveLock };
```

Comment:
Do we really need `AccessShareLock`? As far as I can see, we use `LockColocationId` with `ShareLock` and `ExclsuiveLock`. Can't we stick to those?  Concurrent `EnsureReferenceTablesExistOnAllNodesExtended` should be fine anyway as long as we don't need to replicate placements.

Reply:
Actually a `NoLock` would be enough for the first run I believe. We should _only_ guard against a future colocation group drop during this check, as per third bulletpoint of the PR description.

There is no reason to lock, either moves of shard groups, nor adding removing tables from the colocation group during the first check run. If we would lock any of the other locks we currently use we would want to temporarily block table membership changes here, but that would make create_reference_table be self conflicting.

The actually correct behaviour should have been to remove `EnsureReferenceTablesExistOnAllNodes` from the routines to create a new reference table. However, while implementing that I realized that such a refactor is too intrusive to backport to 10.2.

Reply:
I don't see a downside of acquiring `ShareLock `, I think it is a good practice to hold `ShareLock` while reading, and `ExclusiveLock` while modifying. That sounds more robust for future changes

Reply:
Let me check, but I think it will result in unwanted deadlocks on creating reference tables. Although technically the correcter lock.

Reply:
> I think it would need to drop the sharelock before acquiring the exclusive lock though, to prevent deadlocks.

Ah, that's tricky now I think I can follow why we'd have deadlocks.

Could we maybe always acquire locks on pg-dist-node first? I know, we need a reference table shardid, and calling `CitusTableTypeIdList` currently requires `LockColocationId`.

Could we instead maybe acquire a relation lock to prevent concurrent drop?